### ISSUE TITLE
Dataset for wake words

### ISSUE THREAD

**Issue by farooqkz** [common-voice/common-voice#4733]
**Is your feature request related to a problem? Please describe.**

Wake word detection can be considered an ML problem where our model or solution has to detect if a single word has been spoken or not. There aren't many public wake word datasets available and even if they are, they are behind walls and available for research purposes only.

**Describe the solution you'd like**

We could have a new part of CV, coexisting with the rest where people have to pronounce a single word and validate it. Each person can record a single word multiple times. The current system is not suitable for wake words as single word texts and clips are prohibited. Also I think a single person cannot record the same thing multiple times?

Possibly, users could vote a word to be added to the record-validate process. This way we could have good data few wake words which users like instead tons of words but with data sparingly available for them. Then communities such as those AI assistant ones could tell the users to join CV to vote on a wake word and add it to the DB.

As another alternative, users could see a list of wake words and choose which one to record or validate.

**Describe alternatives you've considered**
Building a new system  can be very much time and energy consuming. Perhaps a new special entry for wake words. It's just like any other text, except it's a single word and should be pronounced as such.

**Additional context**
In both cases, some more data should come with each wake word entry. There should be hints on how to pronounce it. Some words in some languages have different pronounciations based on meaning or even place in a sentence.

**Comment by HarikalarKutusu** [comment]
I also think it is important to have such datasets, especially for edge use in small models like TinyML. A couple remarks here:
- One should extend the idea to commands
- Actually one can have single word sentences in CV - if it is not limited by a language specific validation file.
- Yes, a single sentence can be spoken by a single person only once - to prevent voice bias.
- Additionally, currently a sentence can only be recorded maximum 15 times - by 15 different people.
- The problem is: CV is not a solution for all possible problem. With the voice-AI architectures advanced the other way, so did CV (mainly transcription of talks or assistants etc.).

But, I can do this:
- Build you own server for such data collection
- Remove 15 recording per sentence limit
- Relax the rule where a person can record a sentence single time - but put a max, e.g. 5-10.
- Use sentence domain info as "projects", such as "singlw-word-"1", "washing machine", "hifi-set" etc.

Would be an interesting project...

**Comment by farooqkz** [comment]
Hello. Thank you for your comment. In general CV recommends a sentence rather than a single word. But a wake word is usually a single word or at most just few of them. Expanding this to commands is a good idea. Yeah.

If team members think this is best done as a fork, I will close this issue as invalid/won't fix.

**Comment by HarikalarKutusu** [comment]
@farooqkz, I'm not a team member, but I know CV to an extent. So, lets hear from them first :)

I think it is best done on on a fork due to current workload and limits in the SW. Otherwise, IMO it is a very valid feature request. AFAIK it is done two times, one for numbers, other for COVID words.

It is actually also possible to use a generic model with a language model - finetune for such "hot words" and use intent systems (I've done it for voice-chess), but the model will be unnecessarily large for edge-use for example.

**Comment by farooqkz** [comment]
What are the team members' take on this?

**Comment by jessicarose** [comment]
This is a really interesting perspective, in 2025 we're going to be piloting some datasets that lie outside the traditional Common Voice corpus, starting with a Spontaneous Speech dataset release which introduces a corpus of spontaneous utterances alongside our existing read speech corpus. 

Getting together a wake-word corpus would require a significant amount of team time, so I'm interested in discussing this further to see if there might be any more immediate ways to address this need within the existing contribution structures and datasets, but I'm also just interested in hearing more broadly from you all about what you need from the data!

**Comment by farooqkz** [comment]
Already many wake word datasets exist but in mostly small scale. And some are reserved for academic/research use only.

 - There is hey snips dataset. Has got ~5700 utterances of "Hey snips". Available behind a wall and available for research purposes only.
 - There is Google's Speech commands dataset[1][2] available with recording of about 2 dozens of wake words. But the utterances are limited. And so are the words. The dataset is available under a more open licence. I think a CC variant.
 - There are other datasets available from Qualcomm. Again only for research purposes.
 - There is Mycroft's community dataset. Available under an open licence. But too limited wake words and very few samples per wake word. [3]

As you can see, the common part among all these, is that they are either too limited, too few samples, or come with a too close licencing.

A wake word utterance from part of another sentence may not be best for training wake word detector engines. As some words in some languages have different pronounciations inside a sentence and out. Sometimes it's part of the standard method of the language like Arabic. Sometimes it's just how naturally people tend to speak. For instance the word "home" when prefixed with "at" becomes "adome" as pronounced by some English speakers.

Thus we need the wake word spoken independently or at start of a command "hey computer, up!".

Hoping they won't mind, I'm pinging mycroft members. @chrisemezue @NeonDaniel

I will also ping HomeAssistant Mastodon account.

PS. I've tried the Whisper speech recognition engine from OpenAI on samples of the word "up" from MSWC from MLCommons. Few are recognized as "up".

[1] https://arxiv.org/abs/1804.03209
[2] https://www.tensorflow.org/datasets/catalog/speech_commands
[3] https://github.com/MycroftAI/precise-community-data

**Comment by NeonDaniel** [comment]
> Hoping they won't mind, I'm pinging mycroft members. @chrisemezue @NeonDaniel

Thanks for the tag :slightly_smiling_face: 

> I will also ping HomeAssistant Mastodon account.

@synesthesiam is probably the HomeAssistant dev to look at this.


> This is a really interesting perspective, in 2025 we're going to be piloting some datasets that lie outside the traditional Common Voice corpus, starting with a Spontaneous Speech dataset release which introduces a corpus of spontaneous utterances alongside our existing read speech corpus.

The spontaneous speech dataset sounds like it could be useful as negative training data for WW models on its own. If this is just a collection of specific short phrases, some wake word phrases could potentially be added to that as an easy start to specific wake-word corpora.

> Getting together a wake-word corpus would require a significant amount of team time, so I'm interested in discussing this further to see if there might be any more immediate ways to address this need within the existing contribution structures and datasets, but I'm also just interested in hearing more broadly from you all about what you need from the data!

Collecting data for particular wake words is interesting, but the immediate issue I see is selecting what WWs to collect data for. This [HomeAssistant community WW repository](https://github.com/fwartner/home-assistant-wakewords-collection?tab=readme-ov-file) already has dozens of community submissions (I believe most or all of those are trained on synthetic data). I don't know how to pare down that list, since larger projects will likely want a specific WW and individuals would probably be better off training with their own voice vs a public dataset. A couple very generic wake words (like "hey computer") could be used, though more generic WWs are generally less preferable since they're more likely to come up in regular conversation.

**Comment by HarikalarKutusu** [comment]
> individuals would probably be better off training with their own voice vs a public dataset

I think the best way would be to use a public model and finetune it with the user's voice. If you only use a person's (or family members') voices, it will fail when you catch cold and/or go to the dentists.

**Comment by synesthesiam** [comment]
> I believe most or all of those are trained on synthetic data

I think they all are, which has worked out well for [openWakeWord](https://github.com/dscripka/openWakeWord) because it uses a pre-trained Google embedding model. [microWakeWord](https://github.com/kahrendt/microWakeWord/), on the other hand, performs much better with real data vs. synthetic.

We (the Home Assistant voice team) actually met with some of the people Common Voice today :slightly_smiling_face: 
One of the topics was possibly joining forces on our [Wake Word Collective](https://ohf-voice.github.io/wake-word-collective/). I'd definitely like to see more wake words, especially ones that are language specific. We also need a lot of help getting more balanced datasets in terms of age, high/low vocal register, etc.

**Comment by farooqkz** [comment]
Note that when I said wake words, I didn't mean words meant to wake up assistants only. Imagine you could have an RC car or quadcopter with "forward", "right", "left", "come" and so on.

**Comment by StuartIanNaylor** [comment]
> I think they all are, which has worked out well for [openWakeWord](https://github.com/dscripka/openWakeWord) because it uses a pre-trained Google embedding model. [microWakeWord](https://github.com/kahrendt/microWakeWord/), on the other hand, performs much better with real data vs. synthetic.
> 
> We (the Home Assistant voice team) actually met with some of the people Common Voice today 🙂 One of the topics was possibly joining forces on our [Wake Word Collective](https://ohf-voice.github.io/wake-word-collective/). I'd definitely like to see more wake words, especially ones that are language specific. We also need a lot of help getting more balanced datasets in terms of age, high/low vocal register, etc.

This isn't true about synthetic data, the quality of the synthetic data effects but the dataset construction is paramount.
I have had a brief try with your 'dataset creator' https://github.com/kahrendt/microWakeWord/blob/main/notebooks/basic_training_notebook.ipynb

Its sort of bad in the results it creates and its not Piper, the Vits Piper models are good and have a lot of variation its the model used with piper in that training script that produce very similar sounding synthetic voices where really in terms of intonation variation the only really difference is the 2 gender classes.
Most classification style wakeword use a MFCC 16Khz input which is a quantised spectrogram and really can not contain any synthetic artefacts that you would not be able to detect yourself.

I have trained several wakeword from https://github.com/google-research/google-research/tree/master/kws_streaming using alternative TTS with my own dataset layout and they are all far more accurate than the tflite models you seem to create.

If anyone wants to try use emotivoice as it has '2000' voices of disctinct variation, Coqui xVitts isn't bad as you can get accents by forcing different language models, but they can go on a halucenation warble at times that is interesting if anything else.
Sherpa also do quite easy to install TTS of Vits Piper and Kokoro that also do multi-voice without need to clone.

You do some really weird things in that dataset like adding just RiRs recorded @ 1.5m from https://mcdermottlab.mit.edu/Reverb/IR_Survey.html that includes samples from forrests to subways for a smart_speaker that is to run in the home.

You seem to completely missunderstand RiR and that your dataset needs to be reflective of the rooms with varying voice to mic distances and not supermarkets.

Same goes for your wakeword collective where you advocate for people to walk around a room at any distance and record.
It extemely hard to remove reverberation from mono smaples as the ripple and bounce off surfaces all mix at the mic and depending on how reflective and big the room and voice/mic distance you get distortion that even big tech doesn't know how to remove as there is no reference signal.
That is why smart speakers and far field have multimic and attenuate RiR before wakeword and ASR and why wakeword should not contain reverberation as its easy to add with various utils such as https://github.com/LCAV/pyroomacoustics and near impossible to remove with a ton of compute.

**Comment by StuartIanNaylor** [comment]
You can use synthetic data but additional, real highquality recordings with quality metadata is key.
From Mlcommons spoken word the forced alignment causes huge bias in the dataset to the many small words it captures.
Also secondly as admitted small short words have a high failure rate.
So your creating extremely biased datasets with large proportions of bad labels of extremely varying quality.
Also the English dataset has approx 40% non-native speakers that have very different intonation to english speakers, but the lack of metadata means its impossible to filter and balance a dataset.
From age band, region, native speaker, gender the very basic metadata to create label and class balance is absent.

Openwakeword uses embeddings so its methods are more generalised whilst microWakeWord is a basic image classification model of the incoming audio converted to spectrogram or MFCC. With the right dataset and methods image classification models should be more accurate than generalised embedding ones and with the right dataset the embedding wakeword will likely be better (I don't know Openwakewords methods, just that the dataset creation is the same).
The prob of bad data and overfitting due to low qty's and no metadata is due to what we have and currently a good dataset created with synthetic data can work better the 'real' datasets on offer.

Spectrograms and MFFC are highly reflective of phonetic intonation so if you analyse a language with that 80/20 rule of common words spoken in sentences and then create a smaller subset of phonetically/sylable unique sequence of words you could create dataset, or at least a word set. https://en.wikipedia.org/wiki/Dolch_word_list
Managing to capture high quality has been done but often lacks any granularity of meta-data that often lacks any details on the wildy varying regional accents.
We need more meta-data no-one needs to know where you live or your name, but some very basic details such as region, age band, native speaker but really all we have is language and gender...
If we had datasets of better and more granular metadata the you could train TTS to provide a wider range of voice accents.
Currently ondevice training is often used to provide for the voice variation of accented intonation captured from use.

TTS has come a long way and has models indistinguishable from real voices, but the bias in the sources means they are also biased in output.
Yes you can use them for wakeword dataset but we have a catch22 as they are high quality but lack accent variation whilst real recorded samples in comparison are often lesser quality, that lack displined methodology.

There is far more to just recording samples and throwing them into a collection with no metadata.  
Small ML models are not AI they are just statistical graphs of the dataset you provide and its the old adage of garbage in / garbage out.

Its strange as linguists don't seem to be in the picture and we create datasets based on language ignoring any linguistic science such as https://en.wikipedia.org/wiki/English_phonology.
It works both ways as there are many subsets of english that need identifying metadata but also language trees such as https://en.wikipedia.org/wiki/Germanic_languages with me being a Brit can increase accuracy.
You could likely make a wakeword for this collection of nations https://en.wikipedia.org/wiki/Germanic_languages#/media/File:Germanic_languages_with_dialects_revised.png
That would likely work better if it had a distinct model to https://en.wikipedia.org/wiki/Romance_languages, excluded and vice versa.
That is before we get to the huge differences in European national regional accents that has very little metadata, or sources, due to preference for percieved TV style pronounciation.
From Texas drawl, guttural Scottish to Aussie banter when turned into spectra images of various sorts they have far more difference than synthetic to real equivalents and that is just 'English'.

**Comment by StuartIanNaylor** [comment]
Its worth talking about some points as there are some datasets needed desparately, whilst other datasets are created on the back of missconception, myth and self interest.

Its very simple as an example after reading this thread it prompted me to create a working example of how synthetic data not only works but in many respects is superior to real recordings for the purpose of creating wakewords.
This is because synthetic TTS allows me to use various engines to output phonetic datasets for balanced model classification.

Its good Mike turned up with the claim that synthetic data doesn't work well as the answer is blatant in the dataset creator they have been using for over a year without question or change.
This is a dataset from the Piper dataset creator and you only need to listen once to instantly realise how lacking in prosody / intonation variation it is. Its a basic ML error and the datasets from start are massively overfitted and not reflective of the huge variance of input of use.
https://drive.google.com/file/d/1vFIiW2yeYeBxLxz0XOU5M2ZAKZURPlD4/view?usp=sharing

Here is a 1st stage entire dataset from me that is completely synthetic and likely you will also realise how much more prosody / intonation variance it contains. It is still not good as the base was only 4000 samples / voices that with TTS its still s struggle to get wide ranging prosidy intonation we hear in daily life.
https://drive.google.com/file/d/1LBATLZTRUIotEK4eVi2Cg7sARzgAYbfu/view?usp=sharing

I updated a repo to include the models trained on that very dataset and as you will see if you test those models you will see that basic ML errors are being construed that synthetic data is bad for wakeword creation.
 https://github.com/StuartIanNaylor/wake_word_capture

Synthetic data is superior to real recorded samples purely because it is a clean datum so that when you create a dataset you can be in total control of resultant augmentation. The problem with real data is that the environment and signatures of equipment are unknown and it really hard with mono samples to differentiate what is distortion, what is noise and what is reverberation. So when you add these and the real samples already contain but unknown quantities of distortion, noise  and reverberation you can the control the resultant mix and provide blanced disciplined datasets.

That is basically it in a nut shell and its one of setting a controlled environment which is impossible without rigorous disciplined meta-data collection so that practioners can filter and to create balanced datasets.
The second part is that its impossible to ensure that users will actually adere to any type of discipline.

I explained this to OHV that what they are doing with https://ohf-voice.github.io/wake-word-collective/#ok_nabu is completely flawed. I will save you the read and in its present form will create a totally useless dataset because of zero metadata. You canot pour sample soup into basic classification models with no assurance of balance.
From gender, region, age, room size, distance from mic, the RT60 of the room you have no idea in a bring and by sale that any bias is forming...

I find it staggering no-one involved at HomeAssistant has listened or questioned to how obviously overfitted the initial dataset creation is. Even though my intent is to use many voices of much wider prosidy its still really hard to find TTS that provide good prosody / intonation variation as even some of the latest will clone voice quite well by getting the tonal colour correct but the prosody / intonation seems often stuck to the confines of a narrow language model.

Coqui ⓍTTSv2 https://github.com/coqui-ai/TTSwas the star of the show for me as it allows you to force language settings onto voices of languages they where not trained on and is a gold mine for accented non native english.
Please get that team bat together so they can continue there great work in an effort to stop this obsession with inconsequential word based TTS/ASR in reserach and dev of creating many voice phonetic prosody/intonatin engines.
I used a few TTS and likely if I spent the time I could find many more voices to the dataset but there is a problem out there that they often are just variations of a same language model. That lack the regional variation of national and regional accents that is very apparent if European.
I also used Emoivoice https://github.com/netease-youdao/EmotiVoice that is an extentive 2000+ voices with good prosidy / intonation variation.
Also https://k2-fsa.github.io/sherpa/onnx/tts/index.html# the Kokoro TTS and strangely vits-piper and its strange as I don't know why but that version does have more prosidy / intonation variation and I don't know why as thought they where the same model.

There is only Coqui that has allowed any form of accented no native english that can also work in reverse with the languages it contains.
If the devs continued work extending and increasing the range of langs and voices, but also add regional traits from Texas drawl to Scottish brogue we have very few datasets and near no tools that can synthesize it.
The problem is it needs to be clean so that when augmentation is added we can be sure of resultant levels.

We need something far more rigourous and disciplined than just adhoc recordings under a single piece of metadata of language.
If you are going to create a balanced dataset you need much more than just wakeword as the same voices that record wakeword also need to be included in not wakeword to create the cross entropy so a model can pick distinguishing features which is the !wakeword or often called 'unknown' which would require a huge dataset.
There are such a thing as phonetic pangrams https://clagnut.com/blog/2380/#English_phonetic_pangrams that are short sentences in any language that aims to incorporate all the distinct sounds (phonemes) of a language.
Something like https://github.com/StuartIanNaylor/Dataset-builder as an example because forced alignment is so error prone an on screen prompt of the words in those sentances makes collection hugely less error prone.

Its very easy to add augmentation from RiRs to noise and near imposible without big compute to remove. the latest and best mono single channel blind derverberation uses often uses diffusion models if that gives you any idea how comput intensive that is as opposed how light process such as GPU_RIR and accurately recreate and add to clean samples.
We need clean well recorded samples and this is where I dissagree it can be done not because of common-voice or HA but one of how to provide discipline to any large group of users in donated data...

You can try but walking around a room just providing wakeword lacking any meta-data of any worth to stop bias in your datasets will be ultimately useless...

**Comment by StuartIanNaylor** [comment]
Also here is the collection of piper clean samples before adding noise as I should of said I also run sherpa piper at 0.8 speed as also for me at least the standard intonation is a bit Peruvian marching powder, affected.
https://drive.google.com/file/d/1klsNnmI6fyjL1PKog8RrvOKCI3vV3iZZ/view?usp=sharing so you compare to the piper_generator outout.
I posted a 1st version of the computer-dataset that has 1st stage trained models in https://github.com/StuartIanNaylor/wake_word_capture that also saves capture of the audio it heard. the kw_hit_sensitivity has been turned right down to `default=0.10` so you can also view what might be occasional false positives.
So you can play back samples but also the best dataset for any device model is capture of usage as it includes the signature and artefacts of that devices hardware and algs. Its a direct reflection of its input, which is untrue of multiple recordings oin all types of scenarios and devices.
HA should be in a unique position to allow users to opt-in for automatic data collection and labelling because of the very strong usage patterns of a voice command correctly processed to the user interaction to a voice command processed incorrectly.

True streaming wakeword has the ability to accurated align and capture wakeword with window strides of 30ms or less (20ms in example) whilst still being low load as opposed to rolling window wakeword that stride the input audio to roll the compete input audio to the full non streaming model that sample far less frequently and are not good for capture alignment.

It creates !kw (not wake word) by grabbing a 400,000 word dictionary and via CMUdict and tools creates a sylable and phonetic database so you can run a querry to slect every unique sequence of phonemes to produce a dataset of less than 46000 words.
Excluded from that querry is likeKW which are words with common phonetic sequences to the KW.

This is because all opensource from the oddly named Mycroft precise in the way a simpistic classification structure of unknown, kw & noise create a huge class imbalance, where unknow has a huge array of phonetic features and kw has few.
There are often only 3 classes and there is huge entropy gaps between the classifications of not enough cross entropy to force a model to focus and search for descriminating features as its accuracy will shoot up in an overfitting pattern as it needs very few KW features to get validation right.
Maybe I should of spend a little more thought on selecting the likekw as a way of improving class imbalance and overfitting tendency of kw models.
It doesn't matter as its purely an example that isn't finnished as I am going to trim out current outliers add 2 more classes to the model other the next few days but its likely important to provide each stage example for comparison and evaluation.
So I will add and improve but also to give some points how can can improve and its actually easier with 2 word kw than singular to create a 'sounds' like list containing similar phonetic sequences and likely features and further force the model to find unique KW features.
The notkw label is constructed with all phonetics as often a false positive is a function of softmax where the probility isn't because the KW gets a big hit its because the dividing factors of the other class softmax is very low. low / very low = high.
Hence why soundslike is added to also try and blance that with similar phones.

Simple low computational models are purely a product of thier dataset and image recognition benefits purely from having many classes. Adding classes gives no discernable compute increase so I have given examples, but also explaining why its very important to have alternative words in the same voice as the wakeword or more complex and higher compute models will have to be selected.

The current proposals lack meta-data, accompanying phonetic pangram datasets but discipline and user input will likely be the ultimate fail to the large datasets needed and I have no idea why HA don't capture on device and do local training to embed user voices into models...
I really hope the Coqui team can reform and focus on a phonetic prosidy / intonation engine that provides regional additions to its lang models.

**Comment by farooqkz** [comment]
@StuartIanNaylor Perhaps a little kinder? ;)


I'm also pinging people from OpenVoiceOS to join the discussion. Pinging @goldyfruit @j1nx @JarbasAI. Feel free to ping other people to join the discussion if you think they are interested.

**Comment by StuartIanNaylor** [comment]
:) No not when it will likely be a waste of resources and time. A word dataset needs to be more than just wakeword the user requests for different are endless. Also though its not just wakeword that needs a word dataset as speech enhancement / source seperation / targetted voice seperation all could benefit from have utterance included in the dataset.
They all need clean and dirty samples, whilst clean will remain a major problem.

If you need to add RiR/Noise it is real easy via  augmentation but use correct tools and do it right https://github.com/DavidDiazGuerra/gpuRIR is brilliant you set a room vector W, L, H / speaker vector in that room W, L, H and microphone(s) W,L,H and clean sample set the RT60 and voila.

Its hard not to be critical as after the 1000 blatantly overfitted TTS samples lacking prosody / intonation variation they add hardcoded RiRs that are all fixed at 1.5m from environments as wide as forrests, cathedrals and shopping malls for a smart-speaker that resides in domestic size rooms with many furnishings where RiRs are actually low.

There are limits to how much augmentation you can do where you just increase the entropy of your dataset so that it will become very fuzzy around the edges and the difference between samples can be blurred.
HA say its OK we only apply to loud samples which again adds inbalance as RiRs are not selective on volume, its the room size and reflective nature of surfaces (RT60) that set room reverberation not volume or a threshold.

Augmentation is easy and simple and with a fairly brief search I managed to find TTS that gave me just less than 4000 voices but really that is to small even if it is 4000 of any wakeword and accompanying phonetic pangram or full language dictionary.

When something is broken, its broken. When something is stupid its stupid and not stating so can just be a terribel waste of time for others who might not realise so.
I do have a terrible habit of being very blunt and have already told them what I think that they have gone off in a sulk and blocked me from further comment.
I am not particually bothered as I am not going to sulk and cry and I am still offering the infomation on fixes openly, which there are and wasting time on this supposed wakeword collection waiting for years for user input will still result in many of us needing clean samples so we can balance our datasets via controlled augmentation.

There are now many voice cloning ASR but I am trying to find multivoice 1st with prosody / intonation variation that Coqui does so well as its not just the sample generator script, its just that 1000 are the worst for lack of variation.
Coqui is brilliant for producing accents by mixing voice and langs whilst many others still also just provide voice/lang and also have limits in prosody / intonation.

So lets not waste further time and if people are to record its needs to be clean speech broadcast style that also might have more response that request people walk round rooms whilst randomly saying a wakeword!
This is because a wakeword model is not just a wakeword sample and this is just basic ML that you really don't want to introduce class imbalance, bias or overfitting to your dataset. Also once trained in you can not fine tune out with additional real voices as 'fine' tuning is pretty descriptive and all you are doing is shifting the weights of a overfitted model that will have no effect on that the internal weights are representative of overfitting.

We are now in an era of highly acurate video and audio cloning of a world of digital unreality that is impossible to discern if real or synthetic with forensic tools. When it comes to quantised audio of MFCC there is no longer the resolution in 20x13 vectors that represent 1 second audio.
Maybe because microWakeWord uses an old model https://arxiv.org/pdf/1907.09595 that is from Google but so obselete Google don't even bother to include in https://github.com/google-research/google-research/tree/master/kws_streaming which is the KWS benchmark framework as an accuracy of 78.9% for jaw dropping 7.3M parameters and 565M flops which say in comparison to https://github.com/Qualcomm-AI-research/bcresnet with 96.4 % with only 10k parameters to me is mind blowing they should use something as early in wakeword development as the mix-conv code they have copied that still uses full spectromgram for input and maybe it is more sensitive to synthetic data, but boy is it compute heavy for low accuracy.
The whole industry has shifted to quantise as MFCC to reduce computional cost without loss of resolution for quite some years now and no-one else will be using spectrogram based wakeword especially at the micro-edge buy hey!

If you actually know what is being done and know what it entails, its near impossible not to sound critical and doubt that it has anything to do with the use of synthetic samples.

I have also been training the bcresnet model as can not seem to get the streaming version work from the google-reserch repo so have switched to pytorch and have actually trained a model for the 1st time but being a pytorch noob haven't worked out how to qunatise and convert to onnx or exec-torch as all new to me yet.
All my experience has been tinkering with tensorflow but actually these last 3 years of scratching my head at the voodoo of model training means things are less painful and managed to get a model trained relatively easy.

Again computer  https://drive.google.com/file/d/1NgHVWU3LZSJbGxzy7vBTSqHz_rRqsFie/view?usp=sharing if anyone wants a preview its a bcresnet 1.5 which is just bcresnet1 with spec augment.

The tensorflow tree on esspressif is stagnant in comparison to there Onnx tree but its RkNN that interests me purely out of curiosity as the Rockchips boards have low power hardware VAD and it spurred my curiosity.
I am not really interested in making product or solution its just been an extention of ML curiosity based on my introduction with the fairly terrible Mycroft Precise as it was a swerve ball to me where I have had to learn about the effects of class imbalance, overfitting and all it needs to make much better is to add more classes and squeeze the model to try harder for features to distinguish KW and NotKW.

So longwinded as I am not here to discuss and argue as I will purely provide a model that you can suck and see that synthetic works, but I will give my opinion that style of dataset will further waste much time of others.
Phonetic pangram datasets would be kind of cool especially if you have a feature rich voice of prosody and intonation of many colorfull regional accents especially many poor europeans who are forced to use thier second language.
More Mid West neutral american or British TV english isn't really needed as with have a ton of datasets.
Cool in that they would allow more modern cloning style TTS to provide the variation in prosody / intonation that some 'multi-voice' lack.

@j1nx @JarbasAI are from my very early days at Mycroft and you can use any model even the HA mix-conv if you wish to journey back in the history of wakeword. The model I am training is a CRNN which is a combination of CNN & GRU like precise that is a little more accurate and light than just GRU but near all wake word models of the last decade are somewhere between 90 to near 99% accurate on the Google speech commands benchmark (unless you use an outlier like mixconv).
They will know instantly as when you start to squeeze the KW with 'soundslike' additional classifications the hockey stick of training validation becomes far more a linear ramp and less overfitted.
The proof is just to test the models supplied and updates I will provide, also the dataset was provided to test with whatever model they are using, but include additional classes. Even the benchmark of GSC has far less class imblance as it includes more classes so features are more equally distributed.

If you just collect wakeword alone you are going to condem current models to the class imbalance and overfitting they blatantly display from datasets and training runs and its the flexability and clean source of synthetic that currently means they are better. Whilst those struggling with wakeword models its the creation of the dataset not the source data be it synthetic or real. Some for some reasons do tend to be using very old models that are quite far off current low resource accuracy levels, even if 'better' has been avail as opensource for some time. I guess because tflite4micro support is so limited in what layers it supports as GRU isn't supported just basic CNN or DSCNN will work but LSTM is supported so CRNN needs to be modified to use LSTM than GRU.
It gets worse for what Esspressif have optimised which is very limited as they switched to supporting Onnx in https://github.com/espressif/esp-dl whilst https://github.com/espressif/esp-nn is extremely limited even if current.
If your not using microcontroller then things are much easier and choice of model whatever framework will work on all modern silicon platforms from arm/x86/apple even risc, still though it would seem even though haven't got a quantised model to test the BCresnet model does at least look like its massively ahead in the parameter/accuracy ratio and prob should be a default choice. Whatever model with basic classification your dataset equals accuracy and If your dataset is overfitted, has class imbalance or any form of basic ML mistake, then the trained model will behave so. 
I haven't got mflops ratings for all but parameters is also a pretty good yard stick as it is also for LLM's.

Words are inconsequential to wakeword they are merely phonetic sequence in sylable clumps defined by prosody / intonation that map to spectra in formats such as MFCC and how to balance that in your dataset will be key to how it works in the wild. Too few classes with one such as KW having just few phonetic sylable clumps and a unknown containing all the rest of a 'language' is an instant and huge imbalance.
Then often softmax is used which can also be problematic as with the simple math of softmax your hit might not be because the target class has returned a good hit just that in all other classes the hit is very low.

https://en.wikipedia.org/wiki/Softmax_function

Creating datasets with more classes chosen to distribute what might be the features of your dataset will return much better softmax results than the class imbalance often induced of too few. Explanations are often over complex but softmax is merely your classes features hits divided by all class feature hits to give a supposed probability of 0-1.
So your class can be low and if the other classes contain very few hits you can still return a relatively high score between 0-1... aka false positive.
Also to few too few classes with class imbalance will have much effect on your training run on what your model picks as features and just pouring a sample soup into a single class will not result in accuracy. You can get a feature of a class that is extremely wide because there is a wide gap in cross entropy of your classes.

**Comment by StuartIanNaylor** [comment]
Specifics about dataset used of the for the Kw samples that are still only 3,885 unique voices.

Coqui ⓍTTSv2    870
Emotivoice          1932
Piper (Sherpa)     904
Kokorov1              53
Kokorov1_1          103
Kokoro_en            11
VCKT                      109

A few lost on the way as trimmed and rejected.
More modern 1 shot cloning TTS I didn't use because its equally hard to find sample sources of good variation prosidy / intonation, gender and dialect or further meta-data doesn't seem to exist.
Creating the dataset would of been painfull as transformer based TTS obviously take more compute and because I am unsure what phonetic pangrams would make a good dialect sect my bruteforce method of all unique phoneme sequences in a complete dictionary results in a very large dataset.

The models with only 3,885 unique voices all having some commonality in prosidy / intonation be ause they come from a few older 'multi voice' TTS means my dataset is still overfitted. Just not as overfitted as some.
3.885 was enough as its essential to have a common dataset across all your voice classifications and this is where Phonetic Pangrams of dialect with meta-data will always be rich sources of data as they can represent a complete dialect with a simple sentence "“With tenure, Suzie’d have all the more leisure for yachting, but her publications are no good.”  an example from https://clagnut.com/blog/2380/#English_phonetic_pangrams and if you read the URL it includes with the pangram '(for certain US accents and phonological analyses) '
Its very possible to incase the phonetics of a complete language in a sentence but it can be prosidy / intonation sensitive mainly due to dialects.
Not being a linguist I just brute forced all phonemes from a full language dictionary, which with many unique voices, much more than 3885 voices that are needed my methods would return a ridiculous sized dataset.
In higher education there are likely many linguists who would be able to provide a collection of phonetic pangrams to cover complete languages or at least dialects.
It would be great to have dialects to input into voice technology as resources have all tended to lie on middle neutral ground to the exclusion of a majority.

A wakeword sample on its own is useless as a model is likely to see the difference in dataset as features. The same voice across classes is essential if you want to provide the best and most accurate datasets.
That this is so easily captured by accompaning simle sentences of Phonetic Pangrams would mean data collection would be super simple and little work to record for a complete dataset for any dialect.

Its not just wakeword, its voice filters, source seperation, voice extraction all need clean samples as additive synthesis is accurate and low compute.
It makes zero sense to create just a wakeword dataset when word datasets have many uses and much need.
It doesn't even make anysense for wakeword that are generally post array microphones, speech ehancement and filters.
Whatever HA are doing is a niche need as for general wakeword creation what the propose would be useless as the max levels of noise and reverberation a model will tolerate are quite low.
Hence why big tech has invested heavily in array microphone tech and speech enhancement that sit before wakeword in the audio pipeline.

Phonetic Pangrams are super concise and if I had access to a collection of them I could of used cutting edge cloning tech to create many more voices whilst greatly reducing the dataset processing I had to do.
Wakewords alone are low quality data and those with noise and reverberation recorded in are also because of the dificulties in ascertaining how much is noise and reverberation, whilst clean its just simple to add and both are no different in accuracy with modern tools.
To create a labelled and controlled dataset of phonetic pangram with possible wakeword pairing would massively reduce the work needed to collect samples of varied prosody and intonation of dialects around the world of various demographics. Having complete sets with high quality metadata means you can keep and filter the problems of collection bias out of datasets.

It would be stupid and pointless to collect just wakeword, no metadata or further data for accompanying classes for many applications.
If you where to to start an iniative include more actors for RFC and bring in some liguist experts to help guide what likely would be needed.
There are a ton of uses especialy for modern one shot TSS where phonetic pangrams and wakeword could easilly create wide prosidy/intonation of many language use or specific language cases.

If not synthetic data does work but its sourcing is considerable work hence part of the reason for my arguments and the examples on offer.

There are problems in the historical dataset structure often used for KW and here are my training log directories so with tensorboard you can see the training curve that for any ml practioneer should be a red flag to overfitting bad data.
This is irrespective of synthetic or real.
The 1st is with just a single general likekw added, 2nd likekw is split into 3 based on the ordinal position of the common phoneme to KW, 3rd adds a 1sylable words and a randdom full 1 second sample to further add classifications so that training is forced harder to extract better descriminative features.

https://drive.google.com/file/d/1M-DgGxa9yesKODM2gDTOpq5JIwQShyDS/view?usp=drive_link
https://drive.google.com/file/d/1-kutojc5bC0XsYUiaNWGM46rYvd1s7Qz/view?usp=drive_link


Just use tensorboard to view the training curve tensorboard --logdir .../logs

It takes considerable data to train a wakeword model and a wide range of input is needed to correspond to the wide range of input of use and that is just how basic classification models works.
You can not 'fine tune' out overfitting as iyour fine tuning will always be representitive of the base model wherever you shidt weights to.
If you want to provide for many wakeword and reduce workload then transfer learning of some form would allow you to reuse the effort of an initial model where smaller datasets can be provide via tranfer learning to 'retrain' that model for different wakeword.

Focussing purely on wakeword samples and pouring a sample soup into another classification in a near binary classification structure is basic ml 101 bad practises that without clean controlled samples of the same voices in all voice classifications your dataset is totally imbalanced.

Clean samples well recorded without noise or reverb are needed because it is easy and low compute to add as opposed to remove.
That the need for datasets is not just a need for wakeword in opensource voice tech as voice filters, source seperation, voice extraction and many types of speech enhancement near always require clean samples so you can accurately build controlled and balanced datasets.

Many use librivox style readings of books that are limied in prosidity / intonation of common dialects and lack meta-data.
Also the samples themselves which lack common command sentences of 'Turn on the light' that stops more focussed datasets providing better accuracy with a whole range of opensource voice models.
So that training picks feature differences of the words than artefacts of dissparate recordings its essential 'same' voice exists across classes.
This can be done quite easilly with short concise sentences and commands or wakeword are really not needed as modern cloning TTS provide extremely close and natural prosody / intonation and colour of a reference sample.

We do lack a quality resource of many example dialects reading out a few short Phonetic Pangram sentences, that are rich in metadata and high quality clean recordings.
To garner balanced datasets  of gender, region, age, native language speaker, distance from microphone and sample rate of recording as down sampling is 100% accurate, but 48khz is to provide and the short nature of Phonetic Pangrams means the samples are not all that huge.

Question is what collection Phonetic pangrams provides for all dialects of a single language and consulting linguists and ML practitioners for input would be key to stopping any wasted exorcises of basic ML and audio engineering missconceptions.

Here is the Sqllite word/sylable/phone .db and the simple queries that are purely examples of supporting classes you could use. https://drive.google.com/file/d/195mPQ6NA-Ezq6JkjYL6xFmYx8iCWqmiM/view?usp=sharing
It would be so much simpler to have access to short concise phonetic pangrams of varied prosody / intonation.

**Comment by synesthesiam** [comment]
@farooqkz I'd love to continue the discussion, but this is not possible with Stuart present. He has received numerous warnings, and is no longer welcome in the Home Assistant community.
I'm sure he has some good ideas, but as you can see in this thread alone they are surrounded by excessive noise and an unwillingness to consider other viewpoints.

**Comment by farooqkz** [comment]
> [@farooqkz](https://github.com/farooqkz) I'd love to continue the discussion, but this is not possible with Stuart present. He has received numerous warnings, and is no longer welcome in the Home Assistant community. I'm sure he has some good ideas, but as you can see in this thread alone they are surrounded by excessive noise and an unwillingness to consider other viewpoints.

I'm of the same mind. @jessicarose Is it possible to block them or give a warning or something?

**Comment by moz-kathyreid** [comment]
Thank you @synesthesiam and @farooqkz for bringing this to our attention. My colleague Jessica is on some well-earned leave at the moment, and I will address this via our [Code of Conduct](https://github.com/common-voice/common-voice/blob/main/docs/CODE_OF_CONDUCT.md). 

@farooqkz as the originator of the Issue, I would like to give you control over whether we lock comments on this Issue or not - I will be guided by your wishes.

**Comment by farooqkz** [comment]
Thanks for your quick response. I hope Jessica is enjoying being around Tajiks :D

The problem with @StuartIanNaylor is not their technical points of view. But the way they use words and say things, doesn't lead to a healthy discussion.

Locking the comments blocks any discussion which is also not good. Is it possible to block Stuart from this repository? So we can continue the discussion without him?

Edit: I thought giving him a warning before a block. But I realized I already asked them to change their behavior: https://github.com/common-voice/common-voice/issues/4733#issuecomment-3093215180

It will be also nice if you could hide or delete their comments.

**Comment by StuartIanNaylor** [comment]
> [@farooqkz](https://github.com/farooqkz) I'd love to continue the discussion, but this is not possible with Stuart present. He has received numerous warnings, and is no longer welcome in the Home Assistant community. I'm sure he has some good ideas, but as you can see in this thread alone they are surrounded by excessive noise and an unwillingness to consider other viewpoints.

I have nothing more to say but provide working examples as proof, which leave no argument.

There is no code of conduct broken in anyway apart from to point out the only reason you struggle is the bad methods you are using. Basic ML and audio engineering mistakes leading to bad solutions that have nothing to do with synthetic data.

Your dataset is hugely overfitted but also from Mycroft days all keyword models also have a huge class imbalance that can be simply fixed by adding extra classes that can be easily accomplished by having multiple wakeword indexes in a single model as long as they are relatively phonetically unique to each other.
The same bad methods are just being repeated, misconceptions perpetuated... 
Adding classes causes no additional size or noticeable compute, but gives a classification model more to train against.
Either that and use a more complex model or with simple classification your datset sets outcome.

There is no further discussion than just simple ML fact and that the audio pipeline of a wakeword comes after noise and reverb removal and clean is needed. That also clean has far more uses than just wakeword and that is also just simple fact.
Wakeword models are very simple classification models and you can not throw in noise and reverb above quite low levels as the spectrogram is nolonger representative of source.
This is why big data don't just have simple microphones connected direct to wakeword.
Opensource does have filters and source seperation models to do this but in context of wakeword clean with same voice recordings of supporting dataset is essential or features will not focus on spectragram difference but pick up on the artefacts of different dataset recording in different classes in the same model.
Your proposal in its present form would be a waste of much time and effort so I am unsure what you wish to discuss further, but say what ever you wish. All I will do is post 4 stages of adding classes to a totally synthetic created wakeword model. Highlighting with logs attached obvious overfitting in models and benchmarks where each time more classes are added to increase accuracy, reducing class imbalance and overfitting, as your choice of dataset is like setting a context for training.

I have nothing more to say than provide training logs and finished models purely as an example to how you can improve results and that synthetic data obviously works. I might benchmark them on a one of the libre audio datasets purely to provide further information.

What has been posted is pure fact it is niether kind or cruel, just fact. The viewpoint not being considered is methods being used are common and bad mistakes that some would prefer to quiet than fix, or admit.
I have no interest in what HomeAssistant do but in the wider context of opensource it would be great to provide much neeeded dialect datasets for wake word and all manner of modern ML speech technology and clean is needed.
It would be a waste of time as noise and reverb can be easily added but near impossible to remove especially without leaving artefacts, even working with noise samples as ML-Commons experienced with forced alignment is just so much work with 'dirty' samples and leads to inaccuracies.
There is a wider sphere of use than just HomeAssistant that I hope might be taken into consideration. As said I don't have much more to say than the facts I have provided.

Likely a more knowledgable approach and input from someone like https://speechbrain.github.io/ who have some pretty high end collaborators such as Sherbroke and Cambridge universities.

As for Phonetic Pangrams the example and links I gave are short concise sentences, but purely collect the phonemes of a language. Just great short easy to capture datasets of the much needed dialect datasets that don't exist.
I am not a linguist as prosidy is more than just phonetics, I don't know how big a dataset would need to be to accurately capture dialect prosody, but likely much less than full dictionary language datasets.
I did try myself with several sentences and a KW of my voice several years ago and it created an extremely accurate Wakeword model, but obviously only any use for me. Several magnitudes less data than I have currently used in examples given.

I have updated https://github.com/StuartIanNaylor/wake_word_capture with training runs / models / logs and will do 1 last one.
As said purely an example as my base of just less than 4000 voices still overfits to the wide variety of prosidy you will likely encounter from the wide range of input in the wild.
Coqui ⓍTTSv2 870
Emotivoice 1932
Piper (Sherpa) 904
Kokorov1 53
Kokorov1_1 103
Kokoro_en 11
VCKT 109

Chosen as they quickly allowed collection and the words.db included has the views and base word/syl/phone table to create word lists for TTS.
It will do as an example and when compared to what is provided by the Piper sample generator a quick listen should inform they contain more variation in prosody.
There are a number of much better zero shot clone style TTS that would benefit from having dialect phonetic pangrams as the reference voice for output. I could of gone on a major trawl to create my own prosody dataset collection from the many sources online, its just quite a lot of work and the above for an example was just easier.
I could continue the search for additional TTS and sample collection as often the latest and greatest models can be found on huggingface, but testing and working out scripts for does take time.
So from wakeword, voice filters, source seperation and even ASR modern TTS will outperform low resource Pi style models even for some reason Piper from the onnx sherpa collection outperforms the one used in the training set for micro wake word. A simple listen will tell you what you need to know as yes the TTS from that is not greate for training models and its clearly audible.
Also when creating samples and augmentation its useful to create files names of source and augmentation used as makes post training analysis much easier to stay as audio than preprocessed and pickled datasets.

I will clean up the scripts for TTS automation and post some more repos with references in https://github.com/StuartIanNaylor/wake_word_capture and if you have real samples add them to the dataset as more prosody variance the better, but your really want the same balance in your other classes...
Also there are limits to how much noise and reverberation you can mix into a sample in the example the max of 30% does create a few outliers as the simple classification model nolonger what it recognises as wakeword.
Noise was added as the expectation purely as an example was there would be no early speech enhancement where I find https://github.com/SaneBow/PiDTLN great in use and super light for use on Pizero2 and above.

Even with a PiZero2 and PiDTLN my opinion is that its stupid to use edge methods for a central ASR/LLM the use of edge device should be purely to act as broadcast switch to authoritive wakeword and superior speech enhancement and why I have always been critical of trying to do this at the microedge on microcontrollers when the audio stream is serial. Broadcast unfiltered artefact free likely higherbanwidth audio and use one of the many opensource models that will run on machines capable of running Whisper, especially those runnign local LLMs.
I am not really sure why this did not happen 3 years ago https://community.rhasspy.org/t/thoughts-for-the-future-with-homeassistant-rhasspy/4055/3 or why such a nobrainer should cause so much contention.
There are even invariant adhoc microphone array models that have been available for some years that only need simple wakeword sensors to act as a rooms wide microphone array where the more you add the better the seperation. You only need to focus on 2/3 voice seperation as in normal scences voice1/voice2/noise is the common contention.
Reverb (RT60) in domestic settings is much less of a problem than some seem to think as opposed to noise, that the soft furnishings and room sizes of many surfaces returns quite low RT60 values that are in no way comparitive of supermarkets and forrests and certainly is not loudeness dependent...
ML source seperation does not seem to have the same problems of DSP alaising and models such as https://github.com/yoonsanghyu/FaSNet-TAC-PyTorch do allow the use of wide microphone arrays that limits DSP style models to an approx max of 75mm before alaising becomes problematic.
Again though its same as noise and it has limits what levels you can add for a wakeword dataset as above a threshold the model will no longer recognise as original source.
Speech enhancement should be present earlier in the chain so at least in high levels it should already be filtered.
You can still use great single channel filters such as https://github.com/Rikorose/DeepFilterNet/tree/main/DeepFilterNet as multiple mics can use best wakeword hit to select the best stream.
Much of this can be check by typing "reverb rt60 average room values" and likes into a LLM without need of an audio engineering background, that even simple filters will attenuate as the XMOS is just a simple tflite model running on a multicore microcontroller. A step up to even low cost Pi's can produce much better results to much increased compute capacity. You can use it locally but just don't use it for your central authoritive speech enhancement just use them as low energy sensors to multiplex audio when and where needed.
Great opensource is being ignored unless its refactored and rebranded as own and its my reaction to this where I had decided already and didn't care how my comments where percieved but are also based on fact, but facts some don't want to be heard.

I said my piece on that and that is another issue, here I have purely commented on what I think would be beneficial and what I don't think would be beneficial for opensource. It likely would benefit from knowleable input, especially linguist, before wasting time and resources that is limited in this sector.

**Comment by ftyers** [comment]
Dear all, I'm going to lock this issue as this is not the place for such discussion, please feel free to continue on our [Discourse](https://discourse.mozilla.org/c/voice/239). I would also like to highlight that this kind of use case would be well supported by our upcoming Public API (see [this call](https://common-voice.github.io/common-voice-docs/calls/call_public-api-developer-fund.pdf) for information about our developer fund). We had "target segment" functionality a while ago (3-4 years?), and there are some languages which have a digit corpus + yes/no + "hey firefox!", but there is definitely room for this kind of experimentation.

As an aside @StuartIanNaylor I am not blocking you for now, but I would encourage you to review [Grice's conversational maxims](https://en.wikipedia.org/wiki/Cooperative_principle#Grice's_maxims), particularly that of quantity with a view to improving your future interactions on this medium.
